src/main/java/bio/terra/common/tracing/TracingHook.java

```diff
@@ -0,0 +1,214 @@
+   * Flight with {@link #SUBMISSION_SPAN_CONTEXT_MAP_KEY}.
+   */
+  public static Object serializedCurrentContext() {
+    return encodeContext(Tracing.getTracer().getCurrentSpan().getContext());
+  }
+
+  /** Store the current Span's {@link SpanContext} as the submission Span. */
+  public static void storeCurrentContextAsSubmission(FlightMap inputMap) {
+    inputMap.put(SUBMISSION_SPAN_CONTEXT_MAP_KEY, serializedCurrentContext());
+  }
+
+  /**
+   * Creates a Flight Span for the current run of the Flight. Note that the same Flight may be run
+   * multiple times by Stairway.
+   */
+  @Override
+  public HookAction startFlight(FlightContext context) {
+    SpanContext submissionContext = getOrSubmissionContext(context);
+    // Start the Flight Span. We must remember to close the Flight Span at the end of the Flight's
+    // current run.
+    // We do not currently change the scope to use the Flight Span as the current span. We mostly
+    // only care that the code executed within a step is within the scope of one of our spans.
+    Span flightSpan =
+        Tracing.getTracer()
+            .spanBuilder(FLIGHT_NAME_PREFIX + context.getFlightClassName())
+            .startSpan();
+    flightSpan.addLink(Link.fromSpanContext(submissionContext, Link.Type.PARENT_LINKED_SPAN));
+    flightSpan.putAttribute(
+        "stairway/flightId", AttributeValue.stringAttributeValue(context.getFlightId()));
+    flightSpan.putAttribute(
+        "stairway/flightClass", AttributeValue.stringAttributeValue(context.getFlightClassName()));
+    flightSpans.put(context.getFlightId(), flightSpan);
+    return HookAction.CONTINUE;
+  }
+
+  /** Ends the Flight Span for the current run of the Flight. */
+  @Override
+  public HookAction endFlight(FlightContext context) {
+    Span flightSpan = flightSpans.remove(context.getFlightId());
+    // End the Flight Span for the current run.
+    flightSpan.end();
+    return HookAction.CONTINUE;
+  }
+
+  @Override
+  public Optional<StepHook> stepFactory(FlightContext context) {
+    return Optional.of(new TraceStepHook(flightSpans.get(context.getFlightId())));
+  }
+
+  /** A {@link StepHook} for creating Spans for each Step execution. */
+  private class TraceStepHook implements StepHook {
+    private final Span flightSpan;
+    private Span stepSpan = null;
+    private Scope stepScope = null;
+
+    public TraceStepHook(Span flightSpan) {
+      this.flightSpan = flightSpan;
+    }
+
+    @Override
+    public HookAction startStep(FlightContext flightContext) {
+      // Start the Step Span. We must remember to close the Span at the end of the Step.
+      stepSpan =
+          Tracing.getTracer()
+              .spanBuilderWithExplicitParent(
+                  STEP_NAME_PREFIX + flightContext.getStepClassName(), flightSpan)
+              .startSpan();
+      stepSpan.putAttribute(
+          "stairway/flightId", AttributeValue.stringAttributeValue(flightContext.getFlightId()));
```

Comment:
nit: consider adding stepIndex too?

A serialized version of the input map and working map would be cool too for debugging, but I suppose those may be large and/or contain sensitive values so we shouldn't be persisting them outside of the application db.